Fixed #13883 -- Added optgroups for SelectBox in Admin (updated).

seanhelvey opened the pull request:
#### Trac ticket number
ticket-13883

#### Branch description
The following had already been partially implemented with https://github.com/django/django/pull/15568:
- Modified SelectBox.js so that optgroups are honored. 
- When an item is moved from unselected to selected box, order is maintained by group and then item name. 
- Also added a test.

I forked this [example project](https://github.com/reergymerej/ticket_13883) and made the following updates in [a branch](https://github.com/seanhelvey/ticket_13883), because a few small things were causing errors and confusion in the track discussion:
- Updated ProfileForm with model and fields.
- Used tuple (id, name) instead of (name, name) for choices.
- Changed unicode method to str in Sport model.

Now that the Sport model uses the str method instead of unicode, the following change in comparison was needed: From `if choice_obj == obj:` to `if choice_obj == str(obj):` (in response_add below). Otherwise most of the changes in the original PR and example project had been completed already.

<img width="888" alt="SelectBoxOptGroups" src="https://github.com/user-attachments/assets/68ac09c8-3e4c-4c5a-a176-5e774e54af1b" />

#### Checklist
- [ ] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [ ] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [ ] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

sarahboyce commented:
Thank you for the PR! Would you mind looking into the selenium test failures? :pray:

seanhelvey commented:
> Thank you for the PR! Would you mind looking into the selenium test failures? 🙏

Thanks for letting me know @sarahboyce! Do the Selenium tests run automatically here or do they need to be triggered manually? I hadn't noticed them initially and will get to work. It may take some time, so I can convert this back to draft if it helps. I see that you set the Trac ticket back to needs improvement. 

Leaving notes here for anyone who is interested! :tada: 

Some failures are due to the SelectBox with groups needing to be sorted. It isn't clear to me yet whether we should just change the order of the list elements in the test, modify the tests to account for groups, or create a separate tests for groups.

Other failures are because the User model is in a different app than the Profile model in the tests, whereas the demo app is simpler and they are both in the same app. I'm considering something like source_app_model instead of just source_model so that we can handle this but will have to explore. :hiking_boot:

sarahboyce commented:
> Do the Selenium tests run automatically here or do they need to be triggered manually? I hadn't noticed them initially and will get to work. 

They are triggered when a PR is labeled with "selenium" - so that's why they weren't ran before. They will re-run when updates are pushed to the PR :+1: 

> It may take some time, so I can convert this back to draft if it helps.

I don't mind, there's no rush :slightly_smiling_face:

sarahboyce reviewed (commented):
Thank you for the PR!

I was testing this by changing the User permissions to have option groups:
```diff
--- a/django/contrib/auth/forms.py
+++ b/django/contrib/auth/forms.py
@@ -1,14 +1,19 @@
 import logging
 import unicodedata
+from functools import partial
+from itertools import groupby
+from operator import attrgetter
 
 from django import forms
+from django.contrib.admin.widgets import FilteredSelectMultiple
 from django.contrib.auth import authenticate, get_user_model, password_validation
 from django.contrib.auth.hashers import UNUSABLE_PASSWORD_PREFIX, identify_hasher
-from django.contrib.auth.models import User
+from django.contrib.auth.models import User, Permission
 from django.contrib.auth.tokens import default_token_generator
 from django.contrib.sites.shortcuts import get_current_site
 from django.core.exceptions import ValidationError
 from django.core.mail import EmailMultiAlternatives
+from django.forms.models import ModelChoiceIterator
 from django.template import loader
 from django.utils.encoding import force_bytes
 from django.utils.http import urlsafe_base64_encode
@@ -289,6 +294,30 @@ class UserCreationForm(BaseUserCreationForm):
             return username
 
 
+class GroupedModelChoiceIterator(ModelChoiceIterator):
+    def __init__(self, field, groupby):
+        self.groupby = groupby
+        super().__init__(field)
+
+    def __iter__(self):
+        queryset = self.queryset
+        # Can't use iterator() when queryset uses prefetch_related()
+        if not queryset._prefetch_related_lookups:
+            queryset = queryset.iterator()
+        for group, objs in groupby(queryset, self.groupby):
+            yield (group, [self.choice(obj) for obj in objs])
+
+
+class GroupedModelChoiceField(forms.ModelMultipleChoiceField):
+    def __init__(self, *args, choices_groupby, **kwargs):
+        if isinstance(choices_groupby, str):
+            choices_groupby = attrgetter(choices_groupby)
+        elif not callable(choices_groupby):
+            raise TypeError('choices_groupby must either be a str or a callable accepting a single argument')
+        self.iterator = partial(GroupedModelChoiceIterator, groupby=choices_groupby)
+        super().__init__(*args, **kwargs)
+
+
 class UserChangeForm(forms.ModelForm):
     password = ReadOnlyPasswordHashField(
         label=_("Password"),
@@ -297,6 +326,11 @@ class UserChangeForm(forms.ModelForm):
             "the user’s password."
         ),
     )
+    user_permissions = GroupedModelChoiceField(
+        queryset=Permission.objects.all(),
+        choices_groupby='content_type',
+        widget=FilteredSelectMultiple(verbose_name="User permissions", is_stacked=False),
+    )
 
     class Meta:
         model = User
```

The label ends up in a bit of a strange place:

![Screenshot from 2025-04-17 13-28-52](https://github.com/user-attachments/assets/9f0aa3fd-f5cc-4ba3-8fa0-d295333de97a)


Can you take a look?

seanhelvey commented:
> > Can you take a look?
> 
> I think that alignment issue is [ticket-36053](https://code.djangoproject.com/ticket/36053). When applying the fix from #20075 the issue goes away. The forecast for merging that fix soonish looks good ⛅ , so a rebase should take care of that.
> 
> edit: merged!

Awesome! Rebased and removed the unnecessary blank lines Sarah pointed out. Not sure what to make of the failing visual regression tests.

seanhelvey commented:
I'll keep an eye out for when another rebase is neeeded.

Thanks for suggesting looking into feedback on previous PR here: https://github.com/django/django/pull/15568#pullrequestreview-1490284019.

Re: "User can change the source model and cause a crash" how would they do that? When I add a sport profile with the demo project the user can't edit the url. 
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/8ee2e1a6-0ac5-4c29-ba62-27274a9d3815" />

Re: "Two nested loops with choices are confusing." the data structure is a list of tuples where the value of each tuple is another tuple. What is a cleaner way to process that?

```
> /home/sean/Code/django/django/contrib/admin/options.py(1425)response_add()
-> for optgroup, choices in choices:
(Pdb) choices
[('Team Sports', [(1, 'Duck hunting'), (3, 'Test')]), ('Single Sports', [(2, 'Surfing'), (4, 'Another')])]
```

jacobtylerwalls commented:
Oh the second round of changes already landed in main, you should be good to rebase.

> Re: "Two nested loops with choices are confusing." the data structure is a list of tuples where the value of each tuple is another tuple. What is a cleaner way to process that?

I think the nesting is fine, it's just a naming nitpick on the loop variable (they don't both have to be `choices`).

> Re: "User can change the source model and cause a crash" how would they do that?

I haven't looked in depth, but from Mariusz's screenshot of the debug page, you can see he tampered with `model_source=XXXX` in the URL. It's not the happy path, but I took his point to be that there should be some basic handling of invalid user inputs.

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1398,12 +1401,39 @@ def response_add(self, request, obj, post_url_continue=None):
+            # Find the optgroup for the new item, if available
+            source_model_name = request.POST.get(SOURCE_MODEL_VAR)
+
+            if source_model_name:
```
I can remove this whole `if` block without any test failures.

jacobtylerwalls reviewed (commented):
Hey @seanhelvey, thanks for the updates.

I just got around to checking the test coverage. The most complex part of the patch doesn't have a test.

Once that's in, could you also please add the prior author in the commit message with "Co-authored-by: "?

Adding ".patch" to the commit URL will show the exact signature to use:

https://github.com/django/django/commit/18f0c818abaf4daca42d734d9b28397045437677.patch

Thanks!

seanhelvey commented on django/contrib/admin/options.py:
```diff
@@ -1398,12 +1401,39 @@ def response_add(self, request, obj, post_url_continue=None):
+                try:
+                    source_model = apps.get_model(app_label, model_name)
+                except LookupError:
+                    raise LookupError(f"No installed app/model: {source_model_name}")
```
Added with a few small typo adjustments.

jacobtylerwalls commented on tests/admin_widgets/tests.py:
```diff
@@ -1354,14 +1355,14 @@ def execute_basic_operations(self, mode, field_name):
                 str(self.jenny.id),
                 str(self.john.id),
+                str(self.lisa.id),
+                str(self.peter.id),
```
Can you tell me a little bit about why these are moving?

seanhelvey replied:
I believe SelectBox sorts items by optgroup first, then alphabetically within each group. These test updates reflect that sorted order.

seanhelvey commented on js_tests/admin/SelectBox.test.js:
```diff
@@ -45,3 +45,16 @@ QUnit.test('preserve scroll position', function(assert) {
+    SelectBox.redisplay('id');
+    assert.equal($('#id option').length, 2);
+    assert.equal($('#id optgroup').length, 1);
+});
```
Ok can add. Would it make sense to enable end-of-file-fixer in pre-commit-hooks?

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1398,12 +1401,39 @@ def response_add(self, request, obj, post_url_continue=None):
+                else:
+                    form_class = self.admin_site._registry[source_model].form
+                    if (
+                        hasattr(form_class, "_meta")
```
This still didn't work for me. I found success with `get_form(request)()`:

```py
                    form = self.admin_site._registry[source_model].get_form(request)()
                    field = form.fields[self.opts.verbose_name_plural]
                    if hasattr(field, "choices"):
...

```

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1398,12 +1401,43 @@ def response_add(self, request, obj, post_url_continue=None):
+                                            popup_response["optgroup"] = optgroup_label
+                                            break
+
+            popup_response_data = json.dumps(popup_response)
```
With the GroupAdmin I was testing with [earlier](https://github.com/django/django/pull/18934#discussion_r2673740455), and with the fix from [comment](https://github.com/django/django/pull/18934#discussion_r2705304857), I was able to reach this line with a matched `popup_response["optgroup"]`, but then `dumps()` failed with:

```py
(Pdb) option_value
<ContentType: Myapp | song>

TypeError: Object of type ContentType is not JSON serializable
```

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1398,12 +1401,45 @@ def response_add(self, request, obj, post_url_continue=None):
+                    try:
+                        form = source_admin.get_form(request)()
+                    except (AttributeError, TypeError):
+                        form = None
```
Yeah, I was imagining a test case that did have custom optgroups, along the lines of my example that overrode `formfield_for_manytomany`: https://github.com/django/django/pull/18934#discussion_r2673740455.

jacobtylerwalls replied:
Thanks for the test for the model instances. The part I really wanted to capture was a test for the prior implementation that depended on `_meta` always being available.

In other words, I'm hoping we can get a test that fails if I apply this diff to your branch:
```diff
diff --git a/django/contrib/admin/options.py b/django/contrib/admin/options.py
index 309e35d23e..6fd7585e43 100644
--- a/django/contrib/admin/options.py
+++ b/django/contrib/admin/options.py
@@ -1417,24 +1417,28 @@ class ModelAdmin(BaseModelAdmin):
                     msg = _('The app "%s" could not be found.') % source_model_name
                     self.message_user(request, msg, messages.ERROR)
                 else:
-                    source_admin = self.admin_site._registry[source_model]
-                    form = source_admin.get_form(request)()
-                    if self.opts.verbose_name_plural in form.fields:
-                        field = form.fields[self.opts.verbose_name_plural]
-                        if hasattr(field, "choices"):
-                            for option_value, option_label in field.choices:
-                                # Check if this is an optgroup
-                                # (label is a list/tuple of choices)
-                                if isinstance(option_label, (list, tuple)):
-                                    # It's an optgroup:
-                                    # (group_name, [(value, label), ...])
-                                    optgroup_label = option_value
-                                    for choice_value, choice_display in option_label:
-                                        if choice_display == str(obj):
-                                            popup_response["optgroup"] = str(
-                                                optgroup_label
-                                            )
-                                            break
+                    form_class = self.admin_site._registry[source_model].form
+                    if (
+                        hasattr(form_class, "_meta")
+                        and hasattr(form_class._meta, "fields")
+                        and form_class._meta.fields is not None
+                        and self.opts.verbose_name_plural in form_class._meta.fields
+                    ):
+                        field_choices = (
+                            form_class().fields[self.opts.verbose_name_plural].choices
+                        )
+
+                        for option_value, option_label in field_choices:
+                            # Check if this is an optgroup
+                            # (label is a list/tuple of choices)
+                            if isinstance(option_label, (list, tuple)):
+                                # It's an optgroup:
+                                # (group_name, [(value, label), ...])
+                                optgroup_label = option_value
+                                for choice_value, choice_display in option_label:
+                                    if choice_display == str(obj):
+                                        popup_response["optgroup"] = str(optgroup_label)
+                                        break
 
             popup_response_data = json.dumps(popup_response)
             return TemplateResponse(
```

It should hopefully just be a matter of registering an admin in `tests/admin_views/admin.py` that uses the strategy of overriding a many to many formfield like https://github.com/django/django/pull/18934#discussion_r2673740455. Let me know if this is too hazy.

seanhelvey replied:
I see. Nope the opposite of hazy! Ok that test has been added running through CI now.

jacobtylerwalls commented on django/contrib/admin/static/admin/js/SelectBox.js:
```diff
@@ -1,26 +1,36 @@
-            for (const node of SelectBox.cache[id]) {
-                if (node.displayed) {
-                    const new_option = new Option(node.text, node.value, false, false);
-                    // Shows a tooltip when hovering over the option
```
Preserving this comment might be nice.

jacobtylerwalls commented on docs/releases/6.1.txt:
```diff
@@ -95,6 +95,10 @@ Minor features
 * The admin site login view now redirects authenticated users to the next URL,
   if available, instead of always redirecting to the admin index page.
 
+* Admin popups now preserve :ref:`named groups <field-choices-named-groups>`
```
It's not just the popups, it's the main functionality of the SelectBox.js too, no?

seanhelvey replied:
I am learning with you all good! Updated.

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1398,12 +1401,42 @@ def response_add(self, request, obj, post_url_continue=None):
+                            for option_value, option_label in field.choices:
+                                # Check if this is an optgroup
+                                # (label is a list/tuple of choices)
+                                if isinstance(option_label, (list, tuple)):
```
The doc you linked says any sequence is accepted, so I will liberalize this a little 👍

jacobtylerwalls replied:
I rechecked, and it looks like these get normalized to 2-tuples somewhere no matter how they're initially coded up, so we can leave this as is. 👍

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1398,12 +1401,44 @@ def response_add(self, request, obj, post_url_continue=None):
+                    form = source_admin.get_form(request)()
+                    if self.opts.verbose_name_plural in form.fields:
+                        field = form.fields[self.opts.verbose_name_plural]
+                        if hasattr(field, "choices"):
```
I don't think this attr is ever missing, so I'll deindent.

seanhelvey commented:
> Amazing. The last pushes adjusting sort behavior seem to introduce a new change.
> 
> When use the + popup to create a new item belonging to group already in the "to" box, it doesn't get consolidated, appearing instead at the bottom:
> 
> <img alt="Image" width="543" height="444" src="https://private-user-images.githubusercontent.com/38668450/538719846-cc81d7e0-8f30-4063-81b3-0ed2181cc82e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjkwMjk2MjksIm5iZiI6MTc2OTAyOTMyOSwicGF0aCI6Ii8zODY2ODQ1MC81Mzg3MTk4NDYtY2M4MWQ3ZTAtOGYzMC00MDYzLTgxYjMtMGVkMjE4MWNjODJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTIxVDIxMDIwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmZGFmMTZkNzEzZjA3N2E4OTYwN2JkMGI1OTVjMDU5ZmJjOTNhZmZiZjk2OWJlZjkzY2QwNzUxZGQyOTk1NDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bnHJH1V2nmW7g7vRxqTDwjDUDSGUsAVgaRjgz5CJFb8">
> I would have expected my "ddd" permission to appear in the second group.

I pushed some additional changes here and to the demo project repo so this should be working as expected now, although the build seems to be failing due to unrelated issues.
